### PR TITLE
Implement take for EA mixins

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also _concat_same_type and copy (both based on implementations in #22862)

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/arrays/test_datetimelike.py
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimelike.py]: I think such basic tests are already covered in the base extension tests?

File: pandas/core/arrays/datetimelike.py
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py]: This is not really tested for now, is that correct?

File: pandas/core/arrays/datetimelike.py
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py]: Same comment as in the other PRs, I think we should keep this concept of `_isnan` on the Index (basically my question about it is: for the array classes, what is the advantages of having a `_isnan` instead of simply using `isna()` everywhere?)  

(but since there are a lot of places here where it is used, I am fine with keeping `_isna` itself in here until we de the actual split)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good to me in general (didn't look at the tests in detail), added some comments.

File: pandas/core/arrays/datetimelike.py
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Correct.

File: pandas/tests/arrays/test_datetimelike.py
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py]: No idea.  I'm pretty sure the dtype-specific tests must be non-redundant.

File: pandas/core/arrays/datetimelike.py
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: This is a reasonable view, but one that I'd like to put off dealing with for as long as possible.

The eventual move to composition is going to cause a mismatch between what is and is not cacheable (like `_isnan` is now on the Index subclasses).  Since the EA subclasses actually use `_isnan` in arithmetic/comparison ops, there is a performance hit if we're not careful.  Since I haven't thought of a good solution to this, putting it off is the next best thing.

File: pandas/core/arrays/datetimelike.py
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: Sorry, but I am -1 on moving things out of the PeriodArray PR that cannot stand on its own

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Assuming you're implicitly saying you'd be OK with this if tests were implemented as part of this PR, I think this is an excellent comment.

Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: No, that is not really what I meant :-) 
Because now you added tests for something that would already be tested in the PeriodArray PR. We don't need explicit tests for this method, as it is tested through all existing methods for concat, and it already has explicit tests in the base extension tests.

So I just think this simply belongs in a PR where we actually make PeriodArray an ExtensionArray.

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: > that would already be tested in the PeriodArray PR. 

The benefit of doing things in smaller chunks is that each chunk gets more focused.  e.g. I don't want to throw shade on Tom's PR, but the `take` implementation missed `fill_value` that is a `Period` with non-matching `freq`.  Doing a couple methods at a time lets both authors and reviewers focus more precisely.

And while "too many tests" does have a downside, it really isn't in the top 5 things I'm going to worry about today.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
To be clear / put it a bit more strongly: let's not merge any PR related to periods until we agree on the way forward (smaller like this one or Tom's PR), unless there is an explicit blessing of Tom.

**Comment by jbrockmendel** [comment]
Closing this.  @TomAugspurger i think some of the “take” code/tests may be worth salvaging for the PeriodArray PR.

**Comment by jorisvandenbossche** [comment]
We could also leave this open, so you can rebase once the PeriodArray PR is merged?

**Comment by jorisvandenbossche** [comment]
> but the take implementation missed fill_value that is a Period with non-matching freq

Is that fixed on Tom's PR, or is there an issue for it? Or is that already tested through existing tests?

**Comment by TomAugspurger** [comment]
Fixed locally. Haven't pushed yet.

**Pull request closed without merging**
